CI: Generate RPM diff and auto-publish GitHub Release

ryskn opened the pull request:
Closes #60

Add RPM diff generation and GitHub Release automation to CI

- Detect previous dated image tags with `skopeo` and compare RPM lists
- Generate Markdown changelog (Added / Updated / Removed) per platform
- Publish combined RPM diff as GitHub Release with attached artifacts
- Mark release as pre-release when pushing to testing registries

alexiri commented on .github/workflows/build-test-push.yml:
```diff
@@ -348,6 +347,181 @@ jobs:
+        run: |
+          IMAGE_REPO="${IMAGE_NAMES%%,*}"
+
+          # Fetch tag list and pick latest dated tag (<today): <VMJ><VMIN>-YYYYMMDD
```
Why not just use the current `${VMJ}` tag, which should always be the latest released tag?

unixtech-06 commented on .github/workflows/build-test-push.yml:
```diff
@@ -348,6 +347,181 @@ jobs:
+
+              # Recompute diffs and extract version changes
+              updated_tmp=$(mktemp)
+              awk -F'\t' 'NR==FNR{old[$1]=$2;next}{ if($1 in old && old[$1]!=$2){ print $1"\t"old[$1]" -> "$2 } }' \
```
Thank you for the review
I’ve updated the output arrow to use "->"

unixtech-06 commented on .github/workflows/build-test-push.yml:
```diff
@@ -348,6 +347,181 @@ jobs:
+              delc=$(wc -l < removed.tmp | tr -d ' ')
+
+              echo "" >> "${out}"
+              echo "**Added (${addc})**" >> "${out}"
```
Thanks for the suggestion
I’ve updated the changelog generation so that it now skips any Added / Updated / Removed section when the count is zero

unixtech-06 commented on .github/workflows/build-test-push.yml:
```diff
@@ -348,6 +347,201 @@ jobs:
+                echo ""
+                echo "</details>"
+                echo ""
+              } >> "${out}"
```
That’s a better approach 
switched to a single printf

unixtech-06 commented on .github/workflows/build-test-push.yml:
```diff
@@ -348,6 +347,201 @@ jobs:
+          echo "Checking ${IMAGE_REPO}:${PREV_TAG}"
+          if ! skopeo inspect docker://"${IMAGE_REPO}:${PREV_TAG}" >/dev/null 2>&1; then
+            echo "::error ::Tag not found: ${IMAGE_REPO}:${PREV_TAG}"
+            exit 1
```
updated the step so the workflow won’t fail on the very first run when there’s no previous tag

alexiri commented:
Hi @unixtech-06, please take a look at this conversation: https://chat.almalinux.org/almalinux/pl/yuun4ire5jbibxtid1mfubgzfy

Now would be a great opportunity to standardize on the same SBOM format for all AlmaLinux images!

alexiri commented:
@javihernandez, what do you think, good to go with the SBOM?

javihernandez commented:
> @javihernandez, what do you think, good to go with the SBOM?

Hey, currently, the sbom tools are not part of the cloud images pipelines, it's inclusion in the gcp images pipelines is being finalized as part of https://github.com/AlmaLinux/cloud-images/pull/266.

In any case, yes, I think that we can go the SBOM tools approach. For that @unixtech-06 needs to:
1. Create a PR with the rpm bindings fallback changes in https://github.com/AlmaLinux/cloud-images-sbom-tools
2. Update this PR accordingly. Please check the GCP images PR above for inspiration on how to include the tools into the the container images pipeline(s)

ryskn committed:
feat: SBOM-based changelog generation via OCI referrers

- Replace RPM-based changelog with SBOM (SPDX 2.3 JSON) diff approach
- Generate SBOM per platform using cloud-images-sbom-tools in Fedora container
- Store full SBOM as OCI artifact via ORAS attached to each platform image
- Fetch previous SBOM from OCI referrers for diff comparison
- Publish RPM diff as GitHub Release with Markdown changelog
- Add SBOM metadata label (org.almalinux.sbom) to built images
- Switch registries_testing to quay.io/ryosuke_666
